### ISSUE TITLE
FNX3-16188 ⁃ [Bug] regression: a lot of crashes

### ISSUE THREAD

**Issue by cadeyrn** [mozilla-mobile/fenix#2266]
Fenix crashes a lot since today:

```
2019-05-03 19:37:50.495 16955-16971/org.mozilla.fenix.debug E/WM-WorkerWrapper: Work [ id=bbad165d-6777-4ac0-8444-7e652aad03f7, tags={ Immediate, mozilla.components.feature.sync.WorkManagerSyncWorker, Common } ] failed because it threw an exception/error
    java.util.concurrent.ExecutionException: kotlin.KotlinNullPointerException
        at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
        at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
        at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:289)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: kotlin.KotlinNullPointerException
        at mozilla.components.feature.sync.WorkManagerSyncWorker.doWork(WorkManagerSyncDispatcher.kt:216)
        at androidx.work.CoroutineWorker$startWork$1.invokeSuspend(CoroutineWorker.kt:66)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
```

Happens all the time - when opening websites, closing tabs, typing in the address bar. Sometimes there is no crash, but in many cases Fenix crashes…

### Device information
* Android device: HTC U11 / Android 8.0
* Fenix version: Fenix revision 49ac75c8b20e50da686f6956b7368cd2c6064927

**Comment by cadeyrn** [comment]
Another one:

```
2019-05-03 19:45:12.760 18414-18567/org.mozilla.fenix.debug E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-3
    Process: org.mozilla.fenix.debug, PID: 18414
    android.util.AndroidRuntimeException: Animators may only be run on Looper threads
        at android.animation.AnimatorSet.start(AnimatorSet.java:705)
        at android.animation.AnimatorSet.start(AnimatorSet.java:684)
        at org.mozilla.fenix.components.toolbar.TabCounter.setCountWithAnimation(TabCounter.kt:76)
        at org.mozilla.fenix.components.toolbar.TabCounterToolbarButton.updateCount(TabCounterToolbarButton.kt:55)
        at org.mozilla.fenix.components.toolbar.TabCounterToolbarButton.access$updateCount(TabCounterToolbarButton.kt:19)
        at org.mozilla.fenix.components.toolbar.TabCounterToolbarButton$sessionManagerObserver$1.onSessionRemoved(TabCounterToolbarButton.kt:66)
        at mozilla.components.browser.session.LegacySessionManager$remove$$inlined$synchronized$lambda$1.invoke(LegacySessionManager.kt:291)
        at mozilla.components.browser.session.LegacySessionManager$remove$$inlined$synchronized$lambda$1.invoke(LegacySessionManager.kt:20)
        at mozilla.components.support.base.observer.ObserverRegistry.notifyObservers(ObserverRegistry.kt:99)
        at mozilla.components.browser.session.LegacySessionManager.notifyObservers(Unknown Source:7)
        at mozilla.components.browser.session.LegacySessionManager.remove(LegacySessionManager.kt:291)
        at mozilla.components.browser.session.SessionManager.remove(SessionManager.kt:112)
        at mozilla.components.browser.session.SessionManager.remove$default(SessionManager.kt:111)
        at org.mozilla.fenix.home.HomeFragment$removeTabWithUndo$5.invokeSuspend(HomeFragment.kt:332)
        at org.mozilla.fenix.home.HomeFragment$removeTabWithUndo$5.invoke(Unknown Source:30)
        at org.mozilla.fenix.ext.CoroutineScopeKt$allowUndo$undoJob$1.invokeSuspend(CoroutineScope.kt:24)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
```

**Comment by cadeyrn** [comment]
@ekager Seems also be related to your PR #2246. With the latest build from master (6b349a535a956826e586778fc1202041582814eb) and the changes from #2246 manually removed I no longer see any crashes. After applying PR #2246 again I get the `java.lang.IllegalThreadStateException: Expected thread 2 ("main"), but running on thread 91453 ("DefaultDispatcher-worker-10")` crash again.

**Comment by cadeyrn** [comment]
Fixed by #2281.
